service/efsPort.py

```diff
@@ -0,0 +1,630 @@
+        if not shipHasMedSlots:
+            return None
+
+        filterVal = Item.groupID == getGroup("Propulsion Module").ID
+        propMods = gamedata_session.query(Item).options().filter(filterVal).all()
+        mapPropData = lambda propName: \
+                      next(map(lambda propMod: {"id": propMod.typeID, "powerReq": propMod.attributes["power"].value},
+                               (filter(lambda mod: mod.name == propName, propMods))))
+        mwd5mn = mapPropData("5MN Microwarpdrive II")
+        mwd50mn = mapPropData("50MN Microwarpdrive II")
+        mwd500mn = mapPropData("500MN Microwarpdrive II")
+        mwd50000mn = mapPropData("50000MN Microwarpdrive II")
+        if rigSize == RigSize.SMALL or rigSize is None:
+            propID = mwd5mn["id"] if shipPower > mwd5mn["powerReq"] else None
+        elif rigSize == RigSize.MEDIUM:
+            propID = mwd50mn["id"] if shipPower > mwd50mn["powerReq"] else mwd5mn["id"]
+        elif rigSize == RigSize.LARGE:
+            propID = mwd500mn["id"] if shipPower > mwd500mn["powerReq"] else mwd50mn["id"]
+        elif rigSize == RigSize.CAPITAL:
+            propID = mwd50000mn["id"] if shipPower > mwd50000mn["powerReq"] else mwd500mn["id"]
+
+        if propID is None:
+            return None
+        sFit.appendModule(fitID, propID)
+        sFit.recalc(fit)
+        fit = eos.db.getFit(fitID)
+        mwdPropSpeed = fit.maxSpeed
+        mwdPosition = list(filter(lambda mod: mod.item and mod.item.ID == propID, fit.modules))[0].position
+        sFit.removeModule(fitID, mwdPosition)
+        sFit.recalc(fit)
+        fit = eos.db.getFit(fitID)
+        return mwdPropSpeed
+
+    @staticmethod
+    def getPropData(fit, sFit):
+        fitID = fit.ID
+        propGroupId = getGroup("Propulsion Module").ID
+        propMods = filter(lambda mod: mod.item and mod.item.groupID == propGroupId, fit.modules)
+        activePropWBloomFilter = lambda mod: mod.state > 0 and "signatureRadiusBonus" in mod.item.attributes
+        propWithBloom = next(filter(activePropWBloomFilter, propMods), None)
+        if propWithBloom is not None:
+            oldPropState = propWithBloom.state
+            propWithBloom.state = 0
+            sFit.recalc(fit)
+            fit = eos.db.getFit(fitID)
+            sp = fit.maxSpeed
+            sig = fit.ship.itemModifiedAttributes["signatureRadius"]
+            propWithBloom.state = oldPropState
+            sFit.recalc(fit)
+            fit = eos.db.getFit(fitID)
+            return {"usingMWD": True, "unpropedSpeed": sp, "unpropedSig": sig}
+        return {
+            "usingMWD": False,
+            "unpropedSpeed": fit.maxSpeed,
+            "unpropedSig": fit.ship.itemModifiedAttributes["signatureRadius"]
+        }
+
+    @staticmethod
+    def getOutgoingProjectionData(fit):
+        # This is a subset of module groups capable of projection and a superset of those currently used by efs
+        modGroupNames = [
```

Comment:
Not sure if it's helpful, but you're able to check to see if a module can be projected by using `item.isType("projected")` (I think)(so you would use that in your `for mod in projectedMods:` loop). Maintaining a hardcoded list like this won't be easy. Granted, doing it that way means you might pick up projected modules that EFS can't process, but in that case you might want to implement a checks on the EFS to skip the entry if it's a type you're not expecting.

Reply:
Well groups that aren't handled explicitly won't have a useful set of values anyway.
pyfa considers a lot of stuff to be "projected" that really don't belong here. Missile launchers for example.
It also doesn't cover all the stuff I actually do want like Phenomena Generators and MJDs.
Overall I'm not sure if there's much of a maintenance advantage to using isType("projected") when I'd still need to add any new cases and keep a doc list of what it deals with.

Reply:
Fair points :)